README
NAME
    Date::ICal - Perl extension for ICalendar date objects.

VERSION
    $Revision: 683 $

SYNOPSIS
        use Date::ICal;

        $ical = Date::ICal->new( ical => '19971024T120000' );
        $ical = Date::ICal->new( epoch => time );
        $ical = Date::ICal->new( year => 1964,
            month => 10, day => 16, hour => 16,
            min => 12, sec => 47 );

        $hour = $ical->hour;
        $year = $ical->year;

        $ical_string = $ical->ical;
        $epoch_time = $ical->epoch;

        $ical2 = $ical + $duration;

    (Where $duration is either a duration string, like 'P2W3DT7H9M', or a
    Date::ICal::Duration (qv) object.

        $ical += 'P6DT12H';

        $duration = $ical - $ical2;
        $ical3 = $ical - $duration;

DESCRIPTION
    Date::ICal talks the ICal date format, and is intended to be a base
    class for other date/calendar modules that know about ICal time format
    also.

AUTHOR
    Rich Bowen, and the Reefknot team. Alas, Reefknot is no more. See
    <https://github.com/houseabsolute/DateTime.pm/wiki> for more modern and
    accurate modules.

    Last touched by $Author: michal-josef-spacek $

METHODS
    Date::ICal has the following methods available:

  new
    A new Date::ICal object can be created with any valid ICal string:

        my $ical = Date::ICal->new( ical => '19971024T120000' );
        # will default to the timezone specified in $TZ, see below

    Or with any epoch time:

        my $ical = Date::ICal->new( epoch => time );

    Or, better still, create it with components

        my $date = Date::ICal->new( 
                               day => 25, 
                               month => 10, 
                               year => 1066,
                               hour => 7,
                               min => 15,
                               sec => 47
                               );

    If you call new without any arguments, you'll get a Date::ICal object
    that is set to the time right now.

        my $ical = Date::ICal->new();

    If you already have an object in Date::ICal, or some other subclass
    thereof, you can create a new Date::ICal (or subclass) object using that
    object to start with. This is particularly useful for converting from
    one calendar to another:

       # Direct conversion from Discordian to ISO dates
       my $disco = Date::Discordian->new( disco => '12 Chaos, YOLD 3177' );
       my $iso = Date::ISO->new( $disco );
       print $iso->iso;

    new() handles timezones. It defaults times to UTC (Greenwich Mean Time,
    also called Zulu). If you want to set up a time that's in the US
    "Pacific" timezone, which is GMT-8, use something like:

        my $ical = Date::ICal->new( ical => '19971024T120000',
                                    offset => "-0800");

    Note that as of version 1.44, new() tries to be intelligent about
    figuring out your local time zone. If you enter a time that's not
    *explicitly* in UTC, it looks at the environment variable $TZ, if it
    exists, to determine your local offset. If $TZ isn't set, new() will
    complain.

  ical
        $ical_string = $ical->ical;

    Retrieves, or sets, the date on the object, using any valid ICal
    date/time string. Output is in UTC (ends with a "Z") by default. To get
    output in localtime relative to the current machine, do:

        $ical_string = $ical->ical( localtime => 1 );

    To get output relative to an arbitrary offset, do:

        $ical_string = $ical->ical( offset => '+0545' );

  epoch
        $epoch_time = $ical->epoch;
    
        $ical->epoch( 98687431 );

    Sets, or retrieves, the epoch time represented by the object, if it is
    representable as such. (Dates before 1971 or after 2038 will not have an
    epoch representation.)

    Internals note: The ICal representation of the date is considered the
    only authoritative one. This means that we may need to reconstruct the
    epoch time from the ICal representation if we are not sure that they are
    in synch. We'll need to do clever things to keep track of when the two
    may not be in synch. And, of course, the same will go for any subclasses
    of this class.

  offset_to_seconds
        $seconds_plus_or_minus = offset_to_seconds($offset);

    Changes -0600 to -21600. Not object method, no side-effects.

  offset_from_seconds
        $seconds_plus_or_minus = offset_from_seconds($offset_in_seconds);

    Changes -18000 (seconds) to -0600 (hours, minutes). Not object method,
    no side-effects.

  offset
        $offset = $ical->offset;
   
        # We need tests for these.  
        $ical->offset( '+1100' ); # a number of hours and minutes: UTC+11
        $ical->offset( 0 );       # reset to UTC

    Sets or retrieves the offset from UTC for this time. This allows
    timezone support, assuming you know what your local (or non-local) UTC
    offset is. Defaults to 0.

    Internals note: all times are internally stored in UTC, even though they
    may have some offset information. Offsets are internally stored in
    signed integer seconds.

    BE CAREFUL about using this function on objects that were initialized
    with an offset. If you started an object with:

        my $d = new(ical=>'19700101120000', offset=>'+0100');

    and you then call:

        $d->offset('+0200');

    you'll be saying "Yeah, I know I *said* it was in +0100, but really I
    want it to be in +0200 now and forever." Which may be your intention, if
    you're trying to transpose a whole set of dates to another timezone---
    but you can also do that at the presentation level, with the ical()
    method. Either way will work.

  add
        $self->add( year => 3, month => 2, week => 1, day => 12,
                    hour => 1, min => 34, sec => 59 );
        $date->add( duration => 'P1WT1H1M1S' ); # add 1 wk, 1 hr, 1 min, and 1 sec

    Adds a duration to a Date::ICal object.

    Supported paraters are: duration, eom_mode, year, month, week, day,
    hour, min, sec or seconds.

    'duration' is a ICalendar duration string (see duration_value).

    If a value is undefined or omitted, 1 is assumed:

        $ical->add( 'min' ); # add a minute

    The result will be normalized. That is, the output time will have
    meaningful values, rather than being 48:73 pm on the 34th of
    hexadecember.

    Adding months or years can be done via three different methods,
    specified by the eom_mode parameter, which then applies to all additions
    (or subtractions) of months or years following it in the parameter list.

    The default, eom_mode => 'wrap', means adding months or years that
    result in days beyond the end of the new month will roll over into the
    following month. For instance, adding one year to Feb 29 will result in
    Mar 1.

    If you specify eom_mode => 'limit', the end of the month is never
    crossed. Thus, adding one year to Feb 29, 2000 will result in Feb 28,
    2001. However, adding three more years will result in Feb 28, 2004, not
    Feb 29.

    If you specify eom_mode => 'preserve', the same calculation is done as
    for 'limit' except that if the original date is at the end of the month
    the new date will also be. For instance, adding one month to Feb 29,
    2000 will result in Mar 31, 2000.

    All additions are performed in the order specified. For instance, with
    the default setting of eom_mode => 'wrap', adding one day and one month
    to Feb 29 will result in Apr 1, while adding one month and one day will
    result in Mar 30.

  add_overload
        $date = $date1 + $duration;

    Where $duration is either a duration string, or a Date::ICal::Duration
    object.

        $date += 'P2DT4H7M';

    Adds a duration to a date object. Returns a new object, or, in the case
    of +=, modifies the existing object.

  duration_value
    Given a duration string, this function returns the number of days,
    seconds, and months represented by that duration. In that order. Seems
    odd to me. This should be considered an internal function, and you
    should expect the API to change in the very near future.

  subtract
      $duration = $date1 - $date2;

    Subtract one Date::ICal object from another to give a duration - the
    length of the interval between the two dates. The return value is a
    Date::ICal::Duration object (qv) and allows you to get at each of the
    individual components, or the entire duration string:

        $d = $date1 - $X;

    Note that $X can be any of the following:

    If $X is another Date::ICal object (or subclass thereof) then $d will be
    a Date::ICal::Duration object.

        $week = $d->weeks; # how many weeks apart?
        $days = $d->as_days; # How many days apart?

    If $X is a duration string, or a Date::ICal::Diration object, then $d
    will be an object in the same class as $date1;

        $newdate = $date - $duration;

  clone
        $copy = $date->clone;

    Returns a replica of the date object, including all attributes.

  compare
        $cmp = $date1->compare($date2);

        @dates = sort {$a->compare($b)} @dates;

    Compare two Date::ICal objects. Semantics are compatible with sort;
    returns -1 if $a < $b, 0 if $a == $b, 1 if $a > $b.

  day
        my $day = $date->day;

    Returns the day of the month.

    Day is in the range 1..31

  month
        my $month = $date->month;

    Returns the month of the year.

    Month is returned as a number in the range 1..12

  year
        my $year = $date->year;

    Returns the year.

  jd2greg
        ($year, $month, $day) = jd2greg( $jd );

        Convert number of days on or after Jan 1, 1 CE (Gregorian) to
        gregorian year,month,day.

  greg2jd
        $jd = greg2jd( $year, $month, $day );

        Convert gregorian year,month,day to days on or after Jan 1, 1 CE
        (Gregorian).  Normalization is performed (e.g. month of 28 means
        April two years after given year) for month < 1 or > 12 or day < 1
        or > last day of month.

  days_this_year
      $yday = Date::ICal::days_this_year($day, $month, $year);

    Returns the number of days so far this year. Analogous to the yday
    attribute of gmtime (or localtime) except that it works outside of the
    epoch.

  day_of_week
        my $day_of_week = $date->day_of_week

    Returns the day of week as 0..6 (0 is Sunday, 6 is Saturday).

  hour
        my $hour = $date->hour

    Returns the hour of the day.

    Hour is in the range 0..23

  min
        my $min = $date->min;

    Returns the minute.

    Minute is in the range 0..59

  sec
        my $sec = $date->sec;

    Returns the second.

    Second is in the range 0..60. The value of 60 is (maybe) needed for leap
    seconds. But I'm not sure if we're going to go there.

  julian
      my $jd = $date->jd;

    Returns a listref, containing two elements. The date as a julian day,
    and the time as the number of seconds since midnight. This should not be
    thought of as a real julian day, because it's not. The module is
    internally consistent, and that's enough.

    This method really only is here for compatibility with previous
    versions, as the jd method is now thrown over for plain hash references.

    See the file INTERNALS for more information about this internal format.

TODO
    - add gmtime and localtime methods, perhaps?
    - Fix the INTERNALS file so that it actually reflects reality

INTERNALS
    Please see the file INTERNALS for discussion on the internals.

AUTHOR
    Rich Bowen (DrBacchus) rbowen@rcbowen.com

    And the rest of the Reefknot team. See the source for a full list of
    patch contributors and version-by-version notes.

LICENSE AND COPYRIGHT
    © 2001-2022 Rich Bowen

    © 2022-2023 Michal Josef Špaček

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    datetime@perl.org mailing list

    <https://github.com/houseabsolute/DateTime.pm/wiki>

    Time::Local

    Net::ICal

------------------------------------------------
NAME
    Date::ICal::Duration - durations in iCalendar format, for math purposes.

VERSION
    $Revision: 683 $

SYNOPSIS
        use Date::ICal::Duration;

        $d = Date::ICal::Duration->new( ical => '-P1W3DT2H3M45S' );

        $d = Date::ICal::Duration->new( weeks => 1, 
                                        days => 1,
                                        hours => 6,
                                        minutes => 15,
                                        seconds => 45); 

        # a one hour duration, without other components
        $d = Date::ICal::Duration->new( seconds => "3600"); 

        # Read-only accessors:
        $d->weeks;
        $d->days;
        $d->hours;
        $d->minutes;
        $d->seconds;
        $d->sign;

        # TODO: Resolve sign() discussion from rk-devel and update synopsis.
    
        $d->as_seconds ();   # returns just seconds
        $d->as_elements ();  # returns a hash of elements, like the accessors above
        $d->as_ical();       # returns an iCalendar duration string

DESCRIPTION
    This is a trivial class for representing duration objects, for doing
    math in Date::ICal

AUTHOR
    Rich Bowen, and the Reefknot team. Alas, Reefknot is no more. See
    <https://github.com/houseabsolute/DateTime.pm/wiki> or
    <https://metacpan.org/> for more modern modules.

    Last touched by $Author: rbowen $

METHODS
    Date::ICal::Duration has the following methods available:

  new
    A new Date::ICal::Duration object can be created with an iCalendar
    string :

        my $ical = Date::ICal::Duration->new ( ical => 'P3W2D' );
        # 3 weeks, 2 days, positive direction
        my $ical = Date::ICal::Duration->new ( ical => '-P6H3M30S' );
        # 6 hours, 3 minutes, 30 seconds, negative direction

    Or with a number of seconds:

        my $ical = Date::ICal::Duration->new ( seconds => "3600" );
        # one hour positive

    Or, better still, create it with components

        my $date = Date::ICal::Duration->new ( 
                               weeks => 6, 
                               days => 2, 
                               hours => 7,
                               minutes => 15,
                               seconds => 47,
                               sign => "+"
                               );

    The sign defaults to "+", but "+" and "-" are legal values.

  sign, weeks, days, hours, minutes, seconds
    Read-only accessors for the elements of the object.

  as_seconds
    Returns the duration in raw seconds.

    WARNING -- this folds in the number of days, assuming that they are
    always 86400 seconds long (which is not true twice a year in areas that
    honor daylight savings time). If you're using this for date arithmetic,
    consider using the *add()* method from a Date::ICal object, as this will
    behave better. Otherwise, you might experience some error when working
    with times that are specified in a time zone that observes daylight
    savings time.

  as_days
        $days = $duration->as_days;

    Returns the duration as a number of days. Not to be confused with the
    "days" method, this method returns the total number of days, rather than
    mod'ing out the complete weeks. Thus, if we have a duration of 33 days,
    "weeks" will return 4, "days" will return 5, but "as_days" will return
    33.

    Note that this is a lazy convenience function which is just weeks*7 +
    days.

  as_ical
    Return the duration in an iCalendar format value string (e.g.,
    "PT2H0M0S")

  as_elements
    Returns the duration as a hashref of elements.

INTERNALS
    head2 GENERAL MODEL

    Internally, we store 3 data values: a number of days, a number of
    seconds (anything shorter than a day), and a sign (1 or -1). We are
    assuming that a day is 24 hours for purposes of this module; yes, we

Changes
    * Add L<> in doc to SEE ALSO modules.
    * Cleanup of distribution infrastructure.
    * Remove warning message about deprecation of module.

2023-05-18  skim

    * Add abstract to Makefile.PL.
    * Add author to Makefile.PL.
    * Add LICENSE AND COPYRIGHT section to doc.
    * Add minimal Perl version to Makefile.PL.
    * Add missing license to Makefile.PL.
    * Add use strict and warnings to Makefile.PL.
    * Add use warnings.
    * Unify version number in distribution.

2023-04-04  skim

    * RT#147274: Fix link to datetime project (http://datetime.perl.org is dead).
    * Update README file.

2022-09-07  skim

    * RT#124548: Fix year in timegm(), four digit number now.
    * Fix doc in add() method.

2011-05-10

    * Applies patch from Mark Alway for situation where min, hour, and
    sec are all '00'. However, a reminder that you should be using the
    DateTime modules instead.

2009-08-19 09:23  rbowen

    * No functional changes in this revision, just a move to svn and a
    cleanup of the documentation, removing references to long-defunct
    URLs, and adding references to the datetime.perl.org project, where
    you really should be looking for all of your Perl datetime needs.

2003-01-18 15:11  rbowen

	* MANIFEST (1.10): Added license to manifest

2003-01-18 15:10  rbowen

	* LICENSE (1.1): Put LICENSE in the distribution, since we had
	never made it clear what license we were distributing under.

2003-01-18 15:09  rbowen

	* lib/Date/ICal.pm (1.72): Patch received from Martijn, to make
	days_this_year run faster.

2003-01-13 07:23  rbowen

	* lib/Date/ICal.pm (1.71), t/08offset.t (1.10): Make
	_offset_(from|to)_seconds into offset_(from|to)_seconds. ie. make
	them public functions, rather than hidden ones, since they are
	useful from more places than just inside the module. Requested by
	Martijn.

2003-01-10 15:57  rbowen

	* Makefile.PL (1.20): Had this patch for a long time. Requires a
	particular verion of prereq's for functionality that we are using.

2002-08-31 22:46  rbowen

	* t/: 02normalize.t (1.4), 03components.t (1.3), 04epoch.t (1.3),
	05ical.t (1.2), 06add.t (1.7), 07compare.t (1.2): Change a bunch of
	tests from ok() to is() in order that 'make test' will report more
	useful things on failure.

2002-08-31 22:37  rbowen

	* t/01sanity.t (1.7): Moved from ok() tests to is() tests, in order
	to get more useful failure messages.

2002-06-26 17:21  rbowen

	* lib/Date/ICal.pm (1.70): Documentation bug. new() does not in
	fact accept a 'tz' argument.

2002-02-07 06:33  rbowen

	* lib/Date/ICal/Duration.pm (1.61, Date-ICal-1_69): Bumping up the
	version number on Duration.pm so that CPAN reports useful
	information about what the latest version is.

2002-02-07 06:32  rbowen

	* lib/Date/ICal.pm (1.69), t/08offset.t (1.9) (utags:
	Date-ICal-1_69): Remove skip in test, and actually test the
	localtime functionality.  Adjust seconds the correct direction in
	the ical method.

2002-01-21 14:44  jesse

	* lib/Date/ICal.pm (1.68): Date::ICal->new bitched if you were
	setting an epoch time and had even an offset of "0". That's not
	useful to anyone. it now explicitly checks to make sure an offset
	isn't 0 before bitching.

2001-12-27 21:15  rbowen

	* lib/Date/ICal.pm (1.67), t/12add.t (1.1, Date-ICal-1_69): We are
	now able to use + and += to add durations (either a string or a
	Date::ICal::Duration object) to a Date::ICal object. Added tests
	for same.

2001-12-26 21:37  rbowen

	* t/: 09greg.t (1.2), 10subtract.t (1.5) (utags: Date-ICal-1_69):
	Spacing changes (courtesy of perltidy) in the tests, so that I can
	better read what is going on here. No functional change.

2001-12-26 21:27  rbowen

	* Makefile.PL (1.19, Date-ICal-1_69): Add a vim syntax highlighting
	line to the Change log file so that it looks pretty.

2001-12-26 21:25  rbowen

	* lib/Date/ICal.pm (1.66): No longer need to do strange workaround
	to subtract dates resulting in negative duration. It Just Works
	(tm).

2001-12-26 21:07  rbowen

	* lib/Date/ICal/Duration.pm (1.10), t/11duration.t (1.5,
	Date-ICal-1_69): Permit the creation of Duration objects with
	negative attributes. Note that all attributes have to have the same
	sign. Mixed-sign attributes will result in the entire duration
	being negative. This could stand to be documented more clearly. 
	Also, all accessors have been modified to explicitly return undef
	if that attribute has zero value. I'm not sure if this is desired,
	but it's the way that it is documented, and what the tests expect. 
	Tests added to verify all this stuff.

2001-12-26 20:21  rbowen

	* Makefile.PL (1.18): Add the Duration documentation to the README
	on make dist

2001-12-26 20:15  rbowen

	* lib/Date/ICal/Duration.pm (1.9): Code folds. No functional
	change.

2001-12-26 15:55  rbowen

	* lib/Date/ICal.pm (1.65), t/10subtract.t (1.4): A small
	work-around because D::I::Duration does not correctly handle
	negative arguments yet. See diff 1.7 -> 1.8 of
	Date::ICal::Duration.  When D::I::Duration correctly handles
	negative arguments, this patch should be mostly reversed.  Added
	tests for subtracting dates to get negative duration.

2001-12-26 15:50  rbowen

	* lib/Date/ICal/Duration.pm (1.8): This is step one. Components now
	return signed. And the 'sign' argument is being honored now, rather
	than being discarded. Need to add tests for this stuff. Still does
	not correctly handle negative arguments.

2001-12-20 16:04  rbowen

	* lib/Date/ICal.pm (1.64), t/10subtract.t (1.3): Added the ability
	to subtract durations from a date, either as a Date::ICal::Duration
	object, or as a ICal duration string. Tests to match.

2001-12-20 10:20  rbowen

	* lib/Date/ICal/Duration.pm (1.7): Documentation: The parameter is
	'ical' not 'string' to create an object with an ical duration
	string.

2001-12-19 07:52  rbowen

	* Makefile.PL (1.17), README (1.4): As suggested, generatel README
	at release time, rather than having it in CVS.

2001-12-18 21:24  rbowen


lib/Date/ICal/Duration.pm
package Date::ICal::Duration;

use strict;
use warnings;
use Carp;

use vars qw($VERSION );
$VERSION = '2.'.(qw'$Rev: 683 $')[1];

# Documentation {{{

=encoding utf8

=head1 NAME

Date::ICal::Duration - durations in iCalendar format, for math purposes.

=head1 VERSION

$Revision: 683 $

=head1 SYNOPSIS

    use Date::ICal::Duration;

    $d = Date::ICal::Duration->new( ical => '-P1W3DT2H3M45S' );

    $d = Date::ICal::Duration->new( weeks => 1, 
                                    days => 1,
                                    hours => 6,
                                    minutes => 15,
                                    seconds => 45); 

    # a one hour duration, without other components
    $d = Date::ICal::Duration->new( seconds => "3600"); 

    # Read-only accessors:
    $d->weeks;
    $d->days;
    $d->hours;
    $d->minutes;
    $d->seconds;
    $d->sign;

    # TODO: Resolve sign() discussion from rk-devel and update synopsis.
    
    $d->as_seconds ();   # returns just seconds
    $d->as_elements ();  # returns a hash of elements, like the accessors above
    $d->as_ical();       # returns an iCalendar duration string
    
=head1 DESCRIPTION

This is a trivial class for representing duration objects, for doing math
in Date::ICal

=head1 AUTHOR

Rich Bowen, and the Reefknot team. Alas, Reefknot is no more. See
L<https://github.com/houseabsolute/DateTime.pm/wiki> or L<https://metacpan.org/> for more modern modules.

Last touched by $Author: rbowen $

=head1 METHODS

Date::ICal::Duration has the following methods available:

=head2 new

A new Date::ICal::Duration object can be created with an iCalendar string :

    my $ical = Date::ICal::Duration->new ( ical => 'P3W2D' );
    # 3 weeks, 2 days, positive direction
    my $ical = Date::ICal::Duration->new ( ical => '-P6H3M30S' );
    # 6 hours, 3 minutes, 30 seconds, negative direction
    
Or with a number of seconds:

    my $ical = Date::ICal::Duration->new ( seconds => "3600" );
    # one hour positive

Or, better still, create it with components

    my $date = Date::ICal::Duration->new ( 
                           weeks => 6, 
                           days => 2, 
                           hours => 7,
                           minutes => 15,
                           seconds => 47,
                           sign => "+"
                           );

The sign defaults to "+", but "+" and "-" are legal values. 
                           
=cut

#}}}

#{{{ sub new

sub new {
    my ($class, %args) = @_;
    my $verified = {};
    my $self = {};
    bless $self, $class;

    my $seconds_only = 1;    # keep track of whether we were given length in seconds only
    $seconds_only = 0 unless (defined $args{'seconds'}); 

    # If one of the attributes is negative, then they all must be
    # negative. Otherwise, we're not sure what this means.
    foreach (qw(hours minutes seconds days weeks)) {
        if (defined($args{$_}) )   {
            # make sure this argument is all digits, optional - sign
            if ($args{$_} =~ m/-?[0-9]+$/) { 
                if ($args{$_} < 0) {
                    $args{sign} = '-';
                    $args{$_} = abs($args{$_});
                }
                $verified->{$_} = $args{$_};
                unless ($_ eq 'seconds') {
                    $seconds_only = 0;
                }
            } else {
                carp ("Parameter $_ contains non-numeric value " . $args{$_} . "\n");
            }
        }
    }

    if (defined ($args{sign}) ) {

        # make sure this argument + or -
        if ($args{sign} =~ m/[+-]/) {
            # if so, assign it
            $self->{sign} = ($args{sign} eq "+") ? 1 : -1;
            $verified->{sign} = ($args{sign} eq "+") ? '+' : '-';
        } else {
            carp ("Parameter sign contains a value other than + or - : "
                . $args{sign} . "\n");
        }
        
    }

    # If a number is given, convert it to hours, minutes, and seconds,
    # but *don't* extract days -- we want it to represent an absolute
    # amount of time, regardless of timezone
    if ($seconds_only) { # if we were given an integer time_t
        $self->_set_from_seconds($args{'seconds'});
    } elsif (defined ($args{'ical'}) ) {   
        # A standard duration string
        #warn "setting from ical\n";
        $self->_set_from_ical($args{'ical'});
    } elsif (not $seconds_only) {
        #warn "setting from components";
        #use Data::Dumper; warn Dumper $verified;
        $self->_set_from_components($verified);
    }
    
    return undef unless %args;
   
    return $self;
}

#}}}

# Accessors {{{

=head2 sign, weeks, days, hours, minutes, seconds

Read-only accessors for the elements of the object. 

=cut 

#}}}

# {{{ sub sign 

sub sign {
    my ($self) = @_;
    return $self->{sign};
}

#}}}

# {{{ sub weeks 

sub weeks {
    my ($self) = @_;
    my $w = ${$self->_wd}[0];
    return unless $w;
    return $self->{sign} * $w;
}

#}}}

# {{{ sub days 

sub days {
    my ($self) = @_;
    my $d = ${$self->_wd}[1];
    return unless $d;
    return  $self->{sign} * $d;

} #}}}

#{{{ sub hours 

sub hours {
    my ($self) = @_;
    my $h = ${$self->_hms}[0];
    return unless $h;
    return $self->{sign} * $h;
}

#}}}

# {{{ sub minutes 

sub minutes {
    my ($self) = @_;
    my $m = ${$self->_hms}[1];
    return unless $m;
    return $self->{sign} * $m;
}

#}}}

# {{{ sub seconds 

sub seconds {
    my ($self) = @_;
    my $s = ${$self->_hms}[2];
    return unless $s;
    return $self->{sign} * $s;
}

#}}}

# sub as_seconds {{{

=head2 as_seconds

Returns the duration in raw seconds. 

WARNING -- this folds in the number of days, assuming that they are always 86400
seconds long (which is not true twice a year in areas that honor daylight
savings time).  If you're using this for date arithmetic, consider using the
I<add()> method from a L<Date::ICal> object, as this will behave better.
Otherwise, you might experience some error when working with times that are
specified in a time zone that observes daylight savings time.


=cut 

sub as_seconds {
    my ($self) = @_;

    my $nsecs = $self->{nsecs} || 0;
    my $ndays = $self->{ndays} || 0;
    my $sign  = $self->{sign}  || 1;
    return $sign*($nsecs+($ndays*24*60*60));
}

#}}}

# sub as_days {{{

=head2 as_days

    $days = $duration->as_days;

Returns the duration as a number of days. Not to be confused with the
C<days> method, this method returns the total number of days, rather
than mod'ing out the complete weeks. Thus, if we have a duration of 33
days, C<weeks> will return 4, C<days> will return 5, but C<as_days> will
return 33.

Note that this is a lazy convenience function which is just weeks*7 +
days.

=cut

sub as_days {
    my ($self) = @_;
    my $wd = $self->_wd;
    return $self->{sign} * ( $wd->[0]*7 + $wd->[1] );
}# }}}

#{{{ sub as_ical 

=head2 as_ical

Return the duration in an iCalendar format value string (e.g., "PT2H0M0S")

=cut 

sub as_ical {
    my ($self) = @_;

    my $tpart = '';

    if (my $ar_hms = $self->_hms) {
        $tpart = sprintf('T%dH%dM%dS', @$ar_hms);
    }

    my $ar_wd = $self->_wd();
    
    my $dpart = '';
    if (defined $ar_wd) {
        my ($weeks, $days) = @$ar_wd;
        if ($weeks && $days) {
            $dpart = sprintf('%dW%dD', $weeks, $days);
        } elsif ($weeks) {   # (if days = 0)
            $dpart = sprintf('%dW', $weeks);
        } else {
            $dpart = sprintf('%dD', $days);
        }
    }

    # put a sign in the return value if necessary
    my $value = join('', (($self->{sign} < 0) ? '-' : ''),
                     'P', $dpart, $tpart);

    # remove any zero components from the time string (-P10D0H -> -P10D)
    $value =~ s/(?<=[^\d])0[WDHMS]//g;

    # return either the time value or PT0S (if the time value is zero).
    return (($value !~ /PT?$/) ? $value : 'PT0S');
}

#}}}

#{{{ sub as_elements

=head2 as_elements

Returns the duration as a hashref of elements. 

=cut 

sub as_elements {
    my ($self) = @_;
   
    # get values for all the elements
    my $wd = $self->_wd;
    my $hms = $self->_hms;
   
    my $return = {
        sign => $self->{sign}, 
        weeks => ${$wd}[0],
        days => ${$wd}[1],
        hours => ${$hms}[0],
        minutes => ${$hms}[1],
        seconds => ${$hms}[2],
    };
    return $return;
}

#}}}

# INTERNALS {{{

=head1 INTERNALS

head2 GENERAL MODEL

Internally, we store 3 data values: a number of days, a number of seconds (anything
shorter than a day), and a sign (1 or -1). We are assuming that a day is 24 hours for
purposes of this module; yes, we know that's not completely accurate because of
daylight-savings-time switchovers, but it's mostly correct. Suggestions are welcome.

NOTE: The methods below SHOULD NOT be relied on to stay the same in future versions.

=head2 _set_from_ical ($self, $duration_string)

Converts a RFC2445 DURATION format string to the internal storage format.

=cut

#}}}

# sub _set_from_ical (internal) {{{

sub _set_from_ical {
    my ($self, $str) = @_;

    my $parsed_values = _parse_ical_string($str);
    
    return $self->_set_from_components($parsed_values);
} # }}}

# sub _parse_ical_string (internal) {{{

=head2 _parse_ical_string ($string)

Regular expression for parsing iCalendar into usable values. 

=cut

sub _parse_ical_string {
    my ($str) = @_;
    
    # RFC 2445 section 4.3.6
    #
    # dur-value  = (["+"] / "-") "P" (dur-date / dur-time / dur-week)
    # dur-date   = dur-day [dur-time]
    # dur-time   = "T" (dur-hour / dur-minute / dur-second)
    # dur-week   = 1*DIGIT "W"
    # dur-hour   = 1*DIGIT "H" [dur-minute]
    # dur-minute = 1*DIGIT "M" [dur-second]
    # dur-second = 1*DIGIT "S"
    # dur-day    = 1*DIGIT "D"

    my ($sign_str, $magic, $weeks, $days, $hours, $minutes, $seconds) =
        $str =~ m{
            ([\+\-])?   (?# Sign)
            (P)     (?# 'P' for period? This is our magic character)
            (?:
                (?:(\d+)W)? (?# Weeks)
                (?:(\d+)D)? (?# Days)
            )?
            (?:T        (?# Time prefix)
                (?:(\d+)H)? (?# Hours)
                (?:(\d+)M)? (?# Minutes)
                (?:(\d+)S)? (?# Seconds)
            )?
        }x;

    if (!defined($magic)) {
        carp "Invalid duration: $str";
        return undef;
    }

    # make sure the sign gets set, and turn it into an integer multiplier
    $sign_str ||= "+";
    my $sign = ($sign_str eq "-") ? -1 : 1;
    
    my $return = {};
    $return->{'weeks'} = $weeks;
    $return->{'days'} = $days;
    $return->{'hours'} = $hours;
    $return->{'minutes'} = $minutes;
    $return->{'seconds'} = $seconds;
    $return->{'sign'} = $sign;

    return $return;
} # }}}

# sub _set_from_components (internal) {{{

=head2 _set_from_components ($self, $hashref)

Converts from a hashref to the internal storage format.
The hashref can contain elements "sign", "weeks", "days", "hours", "minutes", "seconds".

=cut

sub _set_from_components {
    my ($self, $args) = @_;

    # Set up some easier-to-read variables
    my ($sign, $weeks, $days, $hours, $minutes, $seconds);
    $sign = $args->{'sign'};
    $weeks = $args->{'weeks'};
    $days = $args->{'days'};
    $hours = $args->{'hours'};
    $minutes = $args->{'minutes'};
    $seconds = $args->{'seconds'};
    
    $self->{sign} = (defined($sign) && $sign eq '-') ? -1 : 1;

    if (defined($weeks) or defined($days)) {
        $self->_wd([$weeks || 0, $days || 0]);
    }

    if (defined($hours) || defined($minutes) || defined($seconds)) {
        $self->_hms([$hours || 0, $minutes || 0, $seconds || 0]);
    }

    return $self;
} # }}}

# sub _set_from_ical (internal) {{{

=head2 _set_from_ical ($self, $num_seconds)

Sets internal data storage properly if we were only given seconds as a parameter.

=cut

sub _set_from_seconds {
    my ($self, $seconds) = @_;
            
    $self->{sign} = (($seconds < 0) ? -1 : 1);
    # find the number of days, if any
    my $ndays = int ($seconds / (24*60*60));
    # now, how many hours/minutes/seconds are there, after
    # days are taken out?
    my $nsecs = $seconds % (24*60*60);
    $self->{ndays} = abs($ndays);
    $self->{nsecs} = abs($nsecs);


    return $self;
} # }}}

# sub _hms (internal) {{{

=head2 $self->_hms();

Return an arrayref to hours, minutes, and second components, or undef
if nsecs is undefined.  If given an arrayref, computes the new nsecs value 
for the duration.  

=cut 

sub _hms {
    my ($self, $hms_arrayref) = @_;

    if (defined($hms_arrayref)) {
        my $new_sec_value = $hms_arrayref->[0]*3600 +
                            $hms_arrayref->[1]*60   + $hms_arrayref->[2];
        $self->{nsecs} = ($new_sec_value);
    } 

    my $nsecs = $self->{nsecs};
    if (defined($nsecs)) {
        my $hours = int($nsecs/3600);
        my $minutes  = int(($nsecs-$hours*3600)/60);
        my $seconds  = $nsecs % 60;
        return [ $hours, $minutes, $seconds ];
    } else {
        print "returning undef\n";
        return undef;
    }
} # }}}

# sub _wd (internal) {{{

=head2 $self->_wd() 

Return an arrayref to weeks and day components, or undef if ndays
is undefined.  If Given an arrayref, computs the new ndays value
for the duration.  

=cut 

sub _wd  {
    my ($self, $wd_arrayref) = @_;

    #print "entering _wd\n";
    
    if (defined($wd_arrayref)) {
        
        my $new_ndays = $wd_arrayref->[0]*7 + $wd_arrayref->[1];
        $self->{ndays} = $new_ndays;
    }
    
    #use Data::Dumper; print Dumper $self->{ndays};
    
    if (defined(my $ndays= $self->{ndays})) {
        my $weeks = int($ndays/7);
        my $days  = $ndays % 7;
        return [ $weeks, $days ];
    } else {
        return undef;
    }
} # }}}

=head1 LICENSE AND COPYRIGHT

© 2001-2022 Rich Bowen

© 2022-2023 Michal Josef Špaček

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

1;
